payment/key.go

```diff
@@ -0,0 +1,104 @@
+package payment
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"encoding/hex"
+	"fmt"
+	"math/rand"
+	"os"
+	"strings"
+	"time"
+)
+
+// AESKey for encrypting secrets
+var AESKey = os.Getenv("ENCRYPTION_KEY")
+var keyLength = 24
+
+// Key represents a merchant's keys to validate skus
+type Key struct {
+	ID                 string     `json:"id" db:"id"`
+	Merchant           string     `json:"merchant" db:"merchant_id"`
+	SecretKey          string     `json:"secretKey"`
+	EncryptedSecretKey string     `json:"-" db:"encrypted_secret_key"`
+	Nonce              string     `json:"-" db:"nonce"`
+	CreatedAt          time.Time  `json:"createdAt" db:"created_at"`
+	Expiry             *time.Time `json:"expiry" db:"expiry"`
+}
+
+// SetSecretKey decrypts the secret key from the database
+func (key *Key) SetSecretKey() {
+	// FIXME call decrypt
+	key.SecretKey = decryptSecretKey(key.EncryptedSecretKey, key.Nonce)
+}
+
+// Taken from https://gist.github.com/kkirsche/e28da6754c39d5e7ea10
+func encryptSecretKey(secretKey string) (string, string) {
+	// The key argument should be the AES key, either 16 or 32 bytes
+	// to select AES-128 or AES-256.
+	key := []byte(AESKey)
+	plaintext := []byte(secretKey)
+
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	// Never use more than 2^32 random nonces with a given key because of the risk of a repeat.
+	nonce := make([]byte, 12)
+	if _, err := rand.Read(nonce); err != nil {
+		panic(err.Error())
+	}
+
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	ciphertext := aesgcm.Seal(nil, nonce, plaintext, nil)
+
+	return fmt.Sprintf("%x", ciphertext), fmt.Sprintf("%x", nonce)
+}
+
+func decryptSecretKey(encryptedSecretKey string, nonceKey string) string {
+	// The key argument should be the AES key, either 16 or 32 bytes
+	// to select AES-128 or AES-256.
+	key := []byte(AESKey)
+	ciphertext, _ := hex.DecodeString(encryptedSecretKey)
```

Comment:
does it make sense to return the error that comes out of these and panic those once in the `SetSecretKey`?

Reply:
+1 we should be propagating the error so we can handle it appropriately in the handler.

Reply:
https://github.com/brave-intl/bat-go/pull/285/commits/f5ca28286a4358dce3ee80d23a3f1c8bdc66e717
https://github.com/brave-intl/bat-go/pull/285/commits/e18d68011e0e937c65a805714650cad416bc3db8